src/definitions.jl

```diff
@@ -329,6 +334,16 @@ function brfft_output_size(sz::Dims{N}, d::Integer, region) where {N}
     return ntuple(i -> i == d1 ? d : sz[i], Val(N))
 end
 
+function output_size(p::Plan)
```

Comment:
Is this type stable? Why not just implement `output_size` for different (sub-)types of `Plan`s instead of relying on a separate trait?

Reply:
The simplest approach to me is to have `AbstractFFTPlan<: Plan`, `AbstractRFFTPlan <: Plan`, and so on. The r2r, dcr stuff would continue to directly subtype `Plan` because there is no special support for them. What do you think?

Reply:
(in such a setting, what would AdjointPlan and ScaledPlan subtype? Would it just be Plan? And they have a type parameter for the underlying plan which we dispatch on?)

Reply:
> Ultimately, what I need from the plan is the projection style (of course, something stronger than the projection style like the FFT function itself is fine too).

I don't think you need that - the easiest would be to just define `output_size(p::Plan) = size(p)` and add specializations for real FFTs such as `output_size(::TestRFFTPlan)` in the tests. You have to be able to dispatch on the plans anyway, even if you work with a `projection_style`, but using an indirect path via `projection_style` which returns something that you can't dispatch on anymore seems more complicated and breaks type inference.

Reply:
I need `projection_style` and not just `output_size` for the implementation of `*` for the transpose plan, unless we do a heuristic based on `output_size`. I totally agree that I should dispatch, but I think that asking downstream implementations to do all of the dispatching themselves would cause them to have a lot of duplicated logic (e.g. forming the scaling) unless we also expose utility functions to make it easier for them. I wanted to make the amount of code needed to be added to downstream packages as little as possible, that's why I thought that we should just ask them to introduce a little more information via types/traits, which we can dispatch on in AbstractFFTs itself

Reply:
Again, I totally agree about switching from the indirect path with `projection_style` to something directly encoded in the type. But I don't think asking downstream plans to provide an implementation of the transpose operation directly makes sense. That's why I'm looking for a design by which we can ask downstream packages to include the necessary extra info in the type, which we can then dispatch on in AbstractFFTs rather than use an if statement.

Reply:
Sure, I understand that. But I don't think the current design of `projection_style` is good. You start from something that you can dispatch on (and hence for which you can define `projection_style`) and instead end up with symbols that you can't dispatch on and the compiler can't use for type inference and code elimination (unless constant propagation kicks in).

Reply:
If you want to go with the projection_style approach you should use dispatchable types instead and use something along the lines of
```julia
struct GenericProjection end
struct RealProjection end
struct RealInverseProjection end
const Projection = Union{GenericProjection,RealProjection,RealInverseProjection} # ensures that there are no other projection types

Projection(p::Plan) = GenericProjection()

output_size(p::Plan) = _output_size(p, Projection(p))
_output_size(p::Plan, ::GenericProjection) = size(p)
...

Base.*(p::AdjointPlan, x::AbstractArray) = _mul(p, x, Projection(p))
_mul(p, x, ::GenericProjection) = ...
```

Reply:
I took a stab at this! Following the approach you suggested nearly completely, with the exception that I did not assume by default that the projection style is none as this would silently fail in many cases.